### PR TITLE
Add token-less re-scrape with request signature

### PR DESCRIPTION
**Pull request by diegoquinteiro**
This PR aims to do trigger a token-less re-scrape of the URLs being updated to trigger a new Instant Article ingestion even if the user does not have a configured app ID/app secret/access token.

- [x] Create key generation/exposure logic
- [x] Gracefully degrade to require a token when OpenSSL lib is not available

### CODE REVIEW

**Comment by diegoquinteiro** [comment]
Hi @jerclarke!

That's one of the changes we're testing to solve the problem, based on your feedback. But we're still rolling out the API side of it and we can only review/land this PR afterwards.

Would you like/be able to help testing it?

Thank you!

**Comment by diegoquinteiro** [comment]
@everton-rosario This PR is now complete and ready for review.

File: facebook-instant-articles.php [hunk 1]
```diff
@@ -563,4 +565,60 @@ function admin_notice__scrape_invalidation_success() {
 	}
 	add_action( 'post_updated', 'invalidate_scrape_on_update', 10, 3 );
 
+	function rescrape_article( $post_id, $post ) {
+		$adapter = new Instant_Articles_Post( $post );
+		$old_slugs = get_post_meta( $post_id, '_wp_old_slug' );
+		if ( $adapter->should_submit_post() ) {
+
```
Comment (everton-rosario) [inline facebook-instant-articles.php]: Remove extra empty line.

File: class-instant-articles-signer.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+  public static function init() {
+    add_action( 'wp', array( 'Instant_Articles_Signer', 'output_public_key' ) );
+  }
+
+  public static function output_public_key() {
+    global $wp;
+    if ( $wp->request == self::PUBLIC_KEY_PATH ) {
+      status_header(200);
+      echo self::get_public_key();
```
Comment (everton-rosario) [inline class-instant-articles-signer.php]: Shouldn't we add proper content-type?

Comment (everton-rosario) [reply class-instant-articles-signer.php]: Check if we shouldn't allow more text into this response. So `the_content` should not have anything else but the key.

File: class-instant-articles-signer.php [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+/**
+ * Class responsible for handling the signature of graph rescrape calls
+ *
+ * @since 4.1
+ */
+class Instant_Articles_Signer {
+
+  const PUBLIC_KEY_OPTION = 'instant-articles-rescrape-public-key';
+  const PRIVATE_KEY_OPTION = 'instant-articles-rescrape-private-key';
+  const PUBLIC_KEY_PATH = '.well-known/graph-api/apikey.pub';
```
Comment (everton-rosario) [inline class-instant-articles-signer.php]: Is this a fixed place? Should we make it by default this path, but allow configuration in the settings page.

Comment (diegoquinteiro) [reply class-instant-articles-signer.php]: This is a fixed place, yes. Will not be configurable.

File: class-instant-articles-signer.php (hunk 2 above)
Comment (diegoquinteiro) [reply class-instant-articles-signer.php]: We `die()` aftwards to ensure nothing else is coming in out the response.

File: facebook-instant-articles.php (hunk 1 above)
Comment (diegoquinteiro) [reply facebook-instant-articles.php]: 💅

**Pull request merged**